auparse/auparse.go

```diff
@@ -66,10 +65,31 @@ type field struct {
 	value string // Parsed and enriched value.
 }
 
-func newField(orig string) *field { return &field{orig: orig, value: orig} }
-func (f *field) Orig() string     { return f.orig }
-func (f *field) Value() string    { return f.value }
-func (f *field) Set(value string) { f.value = value }
+func newField(orig string) field { return field{orig: orig, value: orig} }
+
+type fieldMap map[string]field
+
+func (fm fieldMap) add(key string, f field) {
+	fm[key] = f
+}
+
+func (fm fieldMap) setFieldValue(key, value string) {
+	if f, ok := fm[key]; ok {
+		fm[key] = field{f.orig, value}
+	}
+	return
+}
+
+func (fm fieldMap) find(key string) (field, error) {
+	if f, ok := fm[key]; ok {
+		return f, nil
+	}
+	return field{}, fmt.Errorf("%s key not found", key)
+}
```

Comment:
The only error here is absence, so an `ok bool` would suffice. The absence errors can then be statically defined errors returned by the callers in most cases.

Reply:
```
func (fm fieldMap) find(key string) (field, bool) {
	 f, ok := fm[key]
 	 return f, ok
}
```

Reply:
Apologies, but I don't think I understand your concern in keeping the method as initially proposed in the PR. 
Is it related to performance / readability / or?
The compiler says that only `key` escapes to the heap, and the error stays on the stack.
IIRC I saw the errors also escaping on the heap with the current implementation, but I may be wrong.

The reason for `find` to exist is to check the existence of a key and avoiding repeating the error formatting in every call site.
If `find` would only return the result of map fetch `fm[key]`, it would means that:

* all the callers have to check if `ok` is true _and_
* possibly return an error, that only has the `key` string as difference

This method's signature includes the error to limit the duplication in all the callers at the bottom of the file, IMO they are more readable now, compared with the current version, because the error checking is extracted in `find`.

Reply:
OK. It's a trade-off (though note that the call to `fmt.Errorf` [imposes other costs](https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/fmt/errors.go;l=17-30)); though in the case that the errors are made at the caller site, there is no formatting since the string is already known, and so can be constructed during init, once. I'm not particularly fussed by this since I would expect the error path to be infrequent, but if it weren't, I would prefer the cost to be in the source and not in the runtime.

Reply:
As I say, I'm not that fussed. The way that I had intended it would have had the errors constructed at init and stored as vars, so their allocation would have been amortised to O(0). The code as it is, requires allocation of a `fmt.pp` and parsing of a format string (some of the additional complexity can be seen [here](https://godbolt.org/z/x5Eq43Y6d), but godbolt doesn't show the work that is done in fmt), however the error path is expected to be infrequent, so it's not a huge issue. If it were more likely, I would push back harder.